sdk/cosmos/azure_data_cosmos/src/location_cache.rs

```diff
@@ -0,0 +1,570 @@
+                    endpoints.push(endpoint.clone());
+                } else {
+                    unavailable_endpoints.push(endpoint.clone());
+                }
+            }
+        }
+
+        // Add unavailable endpoints to end of endpoints lists
+        for endpoint in unavailable_endpoints {
+            endpoints.push(endpoint);
+        }
+
+        // If no preferred locations were found, use the default endpoint
+        if endpoints.is_empty() {
+            endpoints.push(default_endpoint.to_string());
+        }
+
+        endpoints
+    }
+
+    fn refresh_stale_endpoints(&mut self) {
+        let mut location_unavailability_info_map =
+            self.location_unavailability_info_map.lock().unwrap();
+
+        // Removes endpoints that have not been checked in the last 5 minutes
+        location_unavailability_info_map.retain(|_, info| {
+            info.last_check_time.elapsed().unwrap_or_default()
+                <= std::time::Duration::from_secs(300)
+        });
+    }
+}
+
+// Tests for location cache
+#[cfg(test)]
+
+mod tests {
+    use super::*;
+    use std::{collections::HashSet, vec};
+
+    fn create_test_data() -> (
+        String,
+        HashMap<String, String>,
+        HashMap<String, String>,
+        Vec<String>,
+    ) {
+        // Setting up test database account data
+        let default_endpoint = "https://default.documents.azure.com".to_string();
+
+        let location_1 = AccountRegion {
+            endpoint: "https://location1.documents.azure.com".to_string(),
+            region: "Location 1".to_string(),
+        };
+        let location_2 = AccountRegion {
+            endpoint: "https://location2.documents.azure.com".to_string(),
+            region: "Location 2".to_string(),
+        };
+        let location_3 = AccountRegion {
+            endpoint: "https://location3.documents.azure.com".to_string(),
+            region: "Location 3".to_string(),
+        };
+        let location_4 = AccountRegion {
+            endpoint: "https://location4.documents.azure.com".to_string(),
+            region: "Location 4".to_string(),
+        };
+        let write_locations = HashMap::from([
+            ("Location 1".to_string(), location_1.endpoint.clone()),
+            ("Location 2".to_string(), location_2.endpoint.clone()),
+        ]);
+
+        let read_locations = HashMap::from([
+            ("Location 1".to_string(), location_1.endpoint.clone()),
+            ("Location 2".to_string(), location_2.endpoint.clone()),
+            ("Location 3".to_string(), location_3.endpoint.clone()),
+            ("Location 4".to_string(), location_4.endpoint.clone()),
+        ]);
+
+        let preferred_locations = vec!["Location 1".to_string(), "Location 2".to_string()];
+
+        (
+            default_endpoint,
+            write_locations,
+            read_locations,
+            preferred_locations,
+        )
+    }
+
+    #[test]
+    fn location_cache_update_test() {
```

Comment:
nit: Remove the `_test` from your test methods. We know they're tests, and they're already in the `tests` module.